doc/Requirements.md

```diff
@@ -0,0 +1,526 @@
+for each "scenario". These "scenario" are then given a priority and a detection predicate.
+Configuration are applied when their associated "scenario" is detected.
+
+"Scenario" are detected through arbitrary criterion provided by the PF host.
+
+## Definition
+
+Configurations: set of values for different parameters.
+A configuration MUST not contain 2 values of the same parameters.
+For example, given a PF with 3 integer parameters A,B,C,
+a configuration can contain
+ - 1 value for (A), (B) or (C)
+ - 2 value for (A,B), (A,C) or (B,C)
+ - 3 value for (A,B,C)
+
+## Configuration
+
+A PF MUST offer configurations as described in the Definition chapter.
+(Note: rule based parameter engine does not manipulate directly values,
+      it applies configuration on the parameters.)
+(Why: This is what the reference implementation does.)
+
+Each configuration MUST be associated with a predicate that condition its eligibility.
+A configuration with a predicate that evaluates to `true` is called an "eligible configuration"
+(Why: This is what the reference implementation does.)
+
+The predicate SHOULD be a "selection criterion rule". See next chapter for a definition.
+(Why: The reference implementation uses a boolean expression based engine.)
+
+## Selection criterion
+
+A selection criterion MUST have one, and only one, state at a given time.
+
+A selection criterion MUST know at construction all possible states.
+(Why: To be able to validate:
+       - rules on start
+       - state changes)
+
+The selection criterion possible states MUST be specifiable by:
+ - Directly a state set (Input -> states == identity)
+   (Note: called exclusive criterion)
+   (Note: An empty set is not allowed as the criterion could not have a state.)
+   (Why: any criterion can be crated with this one.)
+
+The selection criterion possible states SHOULD be specifiable by:
+ - A combination of values
+   (Note: combination in the mathematical sense “ab” -> [“”, “a”, “b”, “ab”])
+   (Note: called inclusive criterion)
+   (Note: Un empty value set is allowed as it's combination -a set containing the empty set- would not be empty.
+          The empty set would be the only possible criteria state.)
+   (Why: The reference implementation supports it.)
+
+### Definitions
+ - Selection criterion rule: function (in the mathematical sense) that MUST given selection criteria return a Boolean.
+
+ - Rule: an Boolean expression of Selection criterion rules
+   (Note: implementation only allows AND and OR combination)
+
+### Criterion change
+
+The api to change criterion values MUST allow atomicity regarding configuration application.
+Ie. It MUST be possible to change multiple criterion values without triggering a configuration
+application.
+(Why: Two criterion might have an excluding state.
+      If configuration application was triggered after each)
+
+### Rules
+
+One MUST always be able to create a selection criterion rule from a given selection criterion
```

Comment:
Not necessarily: it must always be possible to express/write a rule using a selection criterion.